src/v/cluster/producer_state.cc

```diff
@@ -0,0 +1,336 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "producer_state.h"
+
+#include "cluster/logger.h"
+#include "cluster/producer_state_manager.h"
+#include "vassert.h"
+
+namespace cluster {
+
+bool request::operator==(const request& other) const {
+    bool range_match = _first_sequence == other._first_sequence
+                       && _last_sequence == other._last_sequence;
+    // both are in progress or both finished
+    bool status_match = (in_progress() && other.in_progress())
+                        || (!in_progress() && !other.in_progress());
+    bool compare = range_match && status_match;
+    if (compare && !in_progress()) {
+        // both requests have finished
+        // compare the result from promise;
+        auto res = result.get_shared_future().get0();
+        auto res_other = other.result.get_shared_future().get0();
+        bool error_match = (res.has_error() && res_other.has_error())
+                           || (!res.has_error() && !res_other.has_error());
+        compare = compare && error_match;
```

Comment:
is `result` `opearator==` not available ?

Reply:
there isn't one from what I can tell.